include/envoy/event/dispatcher.h

```diff
@@ -260,6 +260,12 @@ class Dispatcher : public DispatcherBase {
    */
   virtual void post(PostCb callback) PURE;
 
+  /**
```

Comment:
While you're at it, could you add this ASSERT to network connection to provide slightly earlier warning about this kind of issue?  Similarly, some thread checks in the grpc async client may also be helpful.

```
--- a/source/common/network/connection_impl.cc
+++ b/source/common/network/connection_impl.cc
@@ -438,6 +438,7 @@ void ConnectionImpl::write(Buffer::Instance& data, bool end_stream) {
 
 void ConnectionImpl::write(Buffer::Instance& data, bool end_stream, bool through_filter_chain) {
   ASSERT(!end_stream || enable_half_close_);
+  ASSERT(dispatcher_.isThreadSafe());
 
   if (write_end_stream_) {
     // It is an API violation to write more data after writing end_stream, but a duplicate
```

Reply:
adding...